src/onewire.rs

```diff
@@ -0,0 +1,135 @@
+//! RMT-based Onewire Implementation
+//!
+//! The Onewire module driver can be used to communicate with onewire (1-Wire)
+//! devices.
+//!
+//! This module is an abstraction around the esp-idf component [onewire_bus](https://components.espressif.com/components/espressif/onewire_bus)
+//! implementation. It is recommended to read the usage of the C API in this [example](https://github.com/espressif/esp-idf/tree/v5.2.2/examples/peripherals/rmt/onewire)
+//!
+//!
+//! This implementation currently supports the one-wire API from the new (v5) esp-idf API.
+//!
+//! The pin this peripheral is attached to must be
+//! externally pulled-up with a 4.7kOhm resistor.
+//!
+//! See the `examples/` folder of this repository for more.
+
+use core::marker::PhantomData;
+use core::ptr;
+
+use esp_idf_sys::*;
+
+use crate::peripheral::Peripheral;
+
+#[derive(Debug)]
+pub struct Device<'b> {
+    _device: onewire_device_t,
+    // _bus: &'b BusDriver<'b>, // not sure how I would hold a reference to this aside from the handle
+    _p: PhantomData<&'b ()>, // holds the lifetime since the device is linked to the lifetime of the bus
+}
+
+impl<'a> Device<'a> {
```

Comment:
OK so how exactly do I send and receive data via the device driver?
I don't see a single method here for doing that?

Reply:
Essentially, a device is just an abstraction for address that we use, similar to the slave ID in I2C, except that it's static. Multiple devices exist on the OneWire bus, and each will require a further driver for their individual registers. I don't provide methods for reading/writing on raw devices because these `read` and `write`  methods are used by the underlying esp-idf C API for specific Devices. 

This does limit implementations to the C API only for now, but at least there is a way to expose them. They are 'extern` in `bindings.rs`. You may notice that none of these functions from the C API use the `onewire_device_t`, communication is still driven through the bus:

```
 pub fn onewire_bus_write_bytes(
        bus: onewire_bus_handle_t,
        tx_data: *const u8,
        tx_data_size: u8,
    ) -> esp_err_t;

 pub fn onewire_bus_read_bytes(
        bus: onewire_bus_handle_t,
        rx_buf: *mut u8,
        rx_buf_size: usize,
    ) -> esp_err_t;

 pub fn onewire_bus_write_bit(bus: onewire_bus_handle_t, tx_bit: u8) -> esp_err_t;

 pub fn onewire_bus_read_bit(bus: onewire_bus_handle_t, rx_bit: *mut u8) -> esp_err_t;

```


https://github.com/DaneSlattery/esp-idf-hal/blob/9443f23fd3bdac6b7023b1d4ae6ef6292ec19dc6/src/onewire.rs#L138  provides a starting point for the R/W. I will work on an example.

Reply:
I do not understand why we should continue wrapping more and more C code - including the C code for the "specific" devices on the bus.

After all, all the additional C code does is to push and pull bits on the wire, likely using as you say read/write bit.

So why don't we just re-implement these additional drivers in Rust instead? There definitely won't be a performance slowdown because of that, because the most important methods - rerding and writing data (a single bit, really?) can be easily exposed and will become the onewire driver RW API.

What I'm trying to understand additionally, is whether there is a scheme which is device-independent as to how the master that speaks to a device "selects" the device to speak to. And how the master "broadcasts" to all devices. Can't imaging this to be device-dependent, because bits and pieces of this onewire thing, like the device "address" (the short and the long one) seem "standardized"?

Reply:
It seems the IO interface of `OWBusDriver` should just be:
- `onewire_bus_write_bytes` - perhaps expose as `write(&mut self, data: &[u8]) -> Result<(), EspError>`?
- `onewire_bus_read_bytes` - perhaps expose as `read(&mut self, buf: &mut [u8]) -> Result<usize, EspError>` where the `usize` is how many bytes were read in the buffer?
- `onewire_bus_reset` - perhaps expose as `reset(&mut self) -> Result<(), EspError>`? (Not that I'm aware what a OW bus "reset" means.)

... and then I expect in the "bytes" above somehow to be an indication of for what device the data is written (for the "write" case) and what device sends us the data (for the "read" case).
I would have expected somehow the read/write methods to have a "device address" of sorts, but I don't see this in the C API. So maybe let me ask again: how do I _address_ a device via the onewire protocol? Can't believe this is device specific?

Reply:
... as for exposing more C API - sure we can do that, but maybe not so ideal. Usually, none of the other protocols (SPI / I2C) work like that. You expose a generic way to read/write data over SPI and I2C and then you write another driver (usually in pure Rust) **on top** of the SPI / I2C protocol. This is e.g. how all the LED display drivers in Rust land work.

Reply:
> how do I address a device via the onewire protocol? Can't believe this is device specific?

From wikipedia:
1. The master starts transmission with a reset pulse, which pulls the wire low for a set time. This resets every slave device on the bus.
2. Every slave device shows that it exists by pulling the bus low for a set time.
3. The basic sequence of data is a reset pulse, followed by an 8-bit command, and then data is sent or received in groups of 8 bits.
4. Each device has a 64-bit serial number which can be used to address it. The LSB of the serial number is common amongst all devices from the same family.
5. There are several standard broadcast commands that all devices will implement. There are also commands to address a particular device. **The master can send a selection command, then the address of a particular device. The following command is then only executed by the addressed device**.

This is quite similar to the I2C protocol. Every device that is attached to the bus has a configured address, and will only respond when it is addressed.

Reply:
Thanks for the pointer. The question is a bit to what extent to model the address of the device explicitly, with an extra parameter to - say - the write command d (not sure how reads work exactly). This is what the i2c driver does actually - https://github.com/esp-rs/esp-idf-hal/blob/master/src/i2c.rs#L276

Reply:
So it turns out that the middle 8 bytes of a message are used for the address when making a request. You can see this in the `rmt_onewire` example. I think transaction logic is a nice to have here, at least for this iteration?

Reply:
I think the latest PR is OK in this regard, so I'll resolve this conversation.